### PR TITLE
[SYCL] Enable gpu testing for aligned allocation functions after L0 bug fix

### PR DESCRIPTION
**Pull request by lbushi25**
I've opened a PR in UR https://github.com/oneapi-src/unified-runtime/pull/1569 which has been merged, that fixes an issue where USM allocation functions are returning non-null pointer values when called with alignment values that are not powers of 2 on L0 devices. In this PR, I've re-enabled gpu testing on an E2E test that was trying to validate this behavior.

### CODE REVIEW

File: sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// REQUIRES: level_zero
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+//==----- alloc_with_invalid_alignment.cpp - SYCL USM allocation test-----==//
```
Comment (uditagarwal97) [inline sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:5]: I don't think it is required to have a license boiler plate in tests, that's the case even with the upstream LLVM.

File: sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===--------------------------------------------------------------------===//
+
+#include <sycl/sycl.hpp>
+
+// Purpose of this test is to verify that when SYCL is backed by L0, the aligned
```
Comment (uditagarwal97) [inline sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:15]: We could simplify the wording. Something like:
"This test verifies that for L0 backend, aligned USM alloc functions returns null_ptr, when called with alignment values that are not a power-of-2. "

Comment (uditagarwal97) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:15]: Also, why just for L0 backend? What happens for opencl backend? If this test is supposed to run on a specific backend, it should have a "REQUIRES: opencl" or "REQUIRES:level_zero" statement

Comment (lbushi25) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:7]: There is a requires level_zero line in the first line of this file.

Comment (uditagarwal97) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:7]: Aahh.. I missed that. 👍🏻

File: sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#include <algorithm>
+#include <cassert>
+#include <sycl/sycl.hpp>
+
+// This test verifies that for L0 backend, aligned USM alloc functions return
+// null_ptr when called with alignment values that are not a power-of-2.
+
+using namespace sycl;
+
+const size_t numBytes = 10;
```
Comment (uditagarwal97) [inline sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:14]: You could inline `numBytes`.

**Review by uditagarwal97 — CHANGES_REQUESTED** [review]
Overall, the PR LGTM. I can approve the changes once the corresponding UR PR gets merged and we accordingly update the UR tag.

**Comment by lbushi25** [comment]
> Overall, the PR LGTM. I can approve the changes once the corresponding UR PR gets merged and we accordingly update the UR tag.

Thanks, will ping you once the UR changes go through.

**Comment by lbushi25** [comment]
UR PR: https://github.com/oneapi-src/unified-runtime/pull/1569

**Comment by kbenzie** [comment]
> Overall, the PR LGTM. I can approve the changes once the corresponding UR PR gets merged and we accordingly update the UR tag.

@uditagarwal97 please rest assured that the [intel/unified-runtime-reviewers](https://github.com/orgs/intel/teams/unified-runtime-reviewers) team, as code owners of `sycl/plugins/unified_runtime/CMakeLists.txt`, will handle this correctly. Requesting changes to block PR's like this actually makes our job harder managing the UR merge backlog by stopping us being able to approve and merge intel/llvm PRs asynchronously once we merge changes to UR main branch. Note we are mostly in Europe, so time zone delays can have a big impact on forward progress.

**Comment by uditagarwal97** [comment]
> > Overall, the PR LGTM. I can approve the changes once the corresponding UR PR gets merged and we accordingly update the UR tag.
> 
> @uditagarwal97 please rest assured that the [intel/unified-runtime-reviewers](https://github.com/orgs/intel/teams/unified-runtime-reviewers) team, as code owners of `sycl/plugins/unified_runtime/CMakeLists.txt`, will handle this correctly. Requesting changes to block PR's like this actually makes our job harder managing the UR merge backlog by stopping us being able to approve and merge intel/llvm PRs asynchronously once we merge changes to UR main branch. Note we are mostly in Europe, so time zone delays can have a big impact on forward progress.

Sure, that makes sense. I'll approve the PR.

**Review by uditagarwal97 — APPROVED** [review]

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 4]
```diff
@@ -94,14 +94,8 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
...
-  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
-  # commit 1e1f577262747d67fcb5c513de4a11bec67ea5db
-  # Merge: dd212f3f 0b405a12
-  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
-  # Date:   Thu May 9 18:00:36 2024 +0100
-  #    Merge pull request #1591 from nrspruit/disable_driver_in_order_lists_default
-  #    [L0] Disable Usage of Driver In order Lists by default
-  set(UNIFIED_RUNTIME_TAG 1e1f577262747d67fcb5c513de4a11bec67ea5db)
+  set(UNIFIED_RUNTIME_REPO "https://github.com/lbushi25/unified-runtime.git")
+  set(UNIFIED_RUNTIME_TAG 74e18f16c0576ba8e20e0017113791335cbfa81c)
```
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:98]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 79b6d698aecd40e44e5872fba15701e10597f694
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Tue May 14 14:22:11 2024 +0100
  #     [HIP] Implement kernel set spec constant query
  #     Handle the `UR_DEVICE_INFO_KERNEL_SET_SPECIALIZATION_CONSTANTS` device
  #     query to report the feature is not supported. This also causes the
  #     `urKernelSetSpecializationConstants` tests to be skipped.
  set(79b6d698aecd40e44e5872fba15701e10597f694)
```

**Review by kbenzie — COMMENTED** [review]
I've now merged https://github.com/oneapi-src/unified-runtime/pull/1569 so please update the UR tag as suggested.

Note however that this PR should only be merged after https://github.com/intel/llvm/pull/13658 in order to keep changes in sync. Once that has merged it will conflict with these changes so it might be better to hold off pulling in the latest sycl branch changes until it has merged.

**Comment by kbenzie** [comment]
@lbushi25 https://github.com/intel/llvm/pull/13658 has now merged, please pull in the latest changes from the sycl branch and resolve the conflict.

**Review by kbenzie — DISMISSED** [review]

**Comment by lbushi25** [comment]
@intel/llvm-gatekeepers friendly ping.

**Comment by lbushi25** [comment]
@intel/llvm-gatekeepers All tests are passing now.

File: sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <algorithm>
+#include <cassert>
+#include <sycl/queue.hpp>
+#include <sycl/usm.hpp>
+
+// This test verifies that for L0 backend, aligned USM alloc functions return
+// null_ptr when called with alignment values that are not a power-of-2.
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: Is this standard behavior? If so, why is it limited to `level_zero`? If not, why do we care?

Comment (aelovikov-intel) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: Also, why does this test belong here and why is there no test in the UR repo?

Comment (lbushi25) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: Yeah, L0 spec says that alignment values should be powers of 2 and SYCL spec says that null should be returned if that is the case. This was created in response to this issue [here](https://github.com/intel/llvm/issues/12638).

For L0 spec see [here](https://spec.oneapi.io/level-zero/latest/core/api.html#zememallocshared) next to the alignment paramater.
For SYCL spec see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_usm_allocations) in the second paragraph.

Comment (lbushi25) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: > Also, why does this test belong here and why is there no test in the UR repo?

I've put here since it is verifying a behaviour mandated by the SYCL spec.  Also, @kbenzie, thoughts on adding a test for this in UR repo? I suppose since this PR does not affect SYCL source code, the test is indirectly testing L0 changes as well but perhaps it could be a good idea to add one there as well.

Comment (aelovikov-intel) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: Please re-write the test to verify what SYCL spec mandates and make it backend-agnostic.

Comment (lbushi25) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: @aelovikov-intel There is some source code to be tested on the SYCL side such as the logic of checking that the L0 allocation functions return invalid codes and then turning them into nullptr from the point of view of a SYCL user and that's what this test is doing. Granted, it's minuscule logic, but I do think it should be tested.

Comment (aelovikov-intel) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: What I'm saying is that `REQUIRES: level_zero` is totally unnecessary. Also, if some backend can align allocation to 17 bytes then it would be valid to return that instead of `nullptr` (although I can't imagine why would that happen). We should also expand this test for big power-of-two requests as well and verify that either required alignment is satisfied or `nullptr` is returned.

Comment (lbushi25) [reply sycl/test-e2e/Regression/alloc_with_invalid_alignment.cpp:11]: > What I'm saying is that `REQUIRES: level_zero` is totally unnecessary. Also, if some backend can align allocation to 17 bytes then it would be valid to return that instead of `nullptr` (although I can't imagine why would that happen). We should also expand this test for big power-of-two requests as well and verify that either required alignment is satisfied or `nullptr` is returned.

Ok, I will make these changes.

**Comment by lbushi25** [comment]
It came to my attention that a very similar test is already in the repo so I just removed an `// UNSUPPORTED: gpu` line to reflect that the bug has been fixed and I will remove this one.

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/test-e2e/USM/align.cpp [hunk 6]
```diff
@@ -35,7 +29,10 @@ template <typename T> void testAlign(sycl::queue &q, unsigned align) {
     return aligned_alloc_host(align, N, args...);
   };
   auto AShared = [&](size_t align, auto... args) {
-    return aligned_alloc_shared(align, N, args...);
+    if (dev.has(aspect::usm_shared_allocations)) {
+      return aligned_alloc_shared(align, N, args...);
+    }
+    return (void*)nullptr;
```
Comment (bader) [inline sycl/test-e2e/USM/align.cpp:35]:
```suggestion
    return (void *)nullptr;
```

File: sycl/test-e2e/USM/align.cpp [hunk 7]
```diff
@@ -48,7 +45,10 @@ template <typename T> void testAlign(sycl::queue &q, unsigned align) {
     return aligned_alloc_host<T>(align, N, args...);
   };
   auto ATShared = [&](size_t align, auto... args) {
-    return aligned_alloc_shared<T>(align, N, args...);
+    if (dev.has(aspect::usm_shared_allocations)) {
+      return aligned_alloc_shared<T>(align, N, args...);
+    }
+    return (T*)nullptr;
```
Comment (bader) [inline sycl/test-e2e/USM/align.cpp:51]:
```suggestion
    return (T *)nullptr;
```

**Review by ldrumm — CHANGES_REQUESTED** [review]

File: sycl/test-e2e/USM/align.cpp [hunk 8]
```diff
@@ -35,7 +29,10 @@ template <typename T> void testAlign(sycl::queue &q, unsigned align) {
     return aligned_alloc_host(align, N, args...);
   };
   auto AShared = [&](size_t align, auto... args) {
-    return aligned_alloc_shared(align, N, args...);
+    if (dev.has(aspect::usm_shared_allocations)) {
+      return aligned_alloc_shared(align, N, args...);
+    }
+    return (void *)nullptr;
```
[the remark this answers is not in the document]
Comment (lbushi25) [reply sycl/test-e2e/USM/align.cpp:35]: Compilation was failing on my machine without the cast. I will apply the change however and see if I get the same error.

[the remark this answers is not in the document]
Comment (lbushi25) [reply sycl/test-e2e/USM/align.cpp:35]: @ldrumm I've reverted back to using the casts because type deduction for the lambda fails otherwise since the types deduced for `nullptr` and the `aligned_alloc_shared` function call are different.

**Comment by lbushi25** [comment]
blocked by https://github.com/oneapi-src/unified-runtime/pull/1672

File: sycl/test-e2e/USM/align.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ldrumm) [reply sycl/test-e2e/USM/align.cpp]: Oh right. Thanks for checking. C++ is wild

**Review by ldrumm — APPROVED** [review]

**Comment by lbushi25** [comment]
Reminder to not merge this until the UR changes go through so that I can update the tag.
